### ISSUE TITLE
Message corruption on the wire with high-frequency (one-way) messages

### ISSUE THREAD

**Issue by andrey-sra** [edenhill/librdkafka#756]
Description
===========

Using Kafka 0.9.0.1 Kafka cluster and recent `master` branch commit for `librdkafka`.  The cluster is rejecting the messages due to negative lengths.  When that happens from a larger number of clients, the cluster can be brought down handling exceptions (OOM).
```
[2016-08-13 06:06:01,052] WARN Unexpected error from /XXX.XXX.XXX.XXX; closing connection (org.apache.kafka.common.network.Selector)
org.apache.kafka.common.network.InvalidReceiveException: Invalid receive (size = -1431655766)
        at org.apache.kafka.common.network.NetworkReceive.readFromReadableChannel(NetworkReceive.java:89)
        at org.apache.kafka.common.network.NetworkReceive.readFrom(NetworkReceive.java:71)
        at org.apache.kafka.common.network.KafkaChannel.receive(KafkaChannel.java:153)
        at org.apache.kafka.common.network.KafkaChannel.read(KafkaChannel.java:134)
        at org.apache.kafka.common.network.Selector.poll(Selector.java:286)
        at kafka.network.Processor.run(SocketServer.scala:413)
        at java.lang.Thread.run(Thread.java:745)
```

How to reproduce
================

1. Install version 2.0+ of tshark (which includes kafka parser).
2. Start a capture in terminal 1 with:
3. `sudo tshark -i eth0 -o 'kafka.tcp.port:9092' -d tcp.port==9092,kafka -f 'dst port 9092' -Y 'kafka.bad_string_length or kafka.bad_bytes_length' -V >> capture.txt` 
4. In terminal 2 start the *modified*
[rdkafka_example.cpp.txt](https://github.com/edenhill/librdkafka/files/416646/rdkafka_example.cpp.txt) example producer:
```./rdkafka_example_cpp -P -t test -d producer -b ...```
Note that the topic test is already heavily written to by other producers (50+) on the network using  the same `librdkafka` library.

*** I am able to provide a pcap traffic capture from the client to the destination port 9092 to whoever is able to take a look at this, which includes 4 Kafka messages with negative length.  Please contact me for URL. ***

Checklist
=========
Please provide the following information:

 - [37c2d08] librdkafka version (release number or git tag):
 - [0.9.1] Apache Kafka version: 
 - [No callbacks+poll] librdkafka client configuration:
 - [ubuntu 14.04] Operating system:
 - [No] Using the legacy Consumer
 - [Yes] Using the high-level KafkaConsumer
 - [cannot repro with debug] Provide logs (with `debug=..` as necessary) from librdkafka
 - [see above] Provide broker log excerpts
 - [X] Critical issue

**Comment by andrey-sra** [comment]
Just to add that even delivery callbacks are enabled, I can still reproduce the issue.  The `tshark` reports corrupted messages:
```                Offset: 0
                Message Size: 341
                Message
                    CRC32: 0xffd14b45
                    Magic Byte: 0
                    .... ..00 = Compression Codec: None (0)
                    Bytes Length: -1
                    Key: <MISSING>
                    Bytes Length: 327
                    Value: bbbbaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa...
                Offset: -6148914691236517206
                Message Size: -1431655766
                Message
                    CRC32: 0xaaaaaaaa
                    Magic Byte: -86
                    .... ..10 = Compression Codec: Snappy (2)
                    Bytes Length: -1431655766
                        [Expert Info (Warn/Malformed): Invalid byte length field]
                            [Invalid byte length field]
                            [Severity level: Warn]
                            [Group: Malformed]
                    Bytes Length: -1431655766
                        [Expert Info (Warn/Malformed): Invalid byte length field]
                            [Invalid byte length field]
                            [Severity level: Warn]
                            [Group: Malformed]
```

**Comment by andrey-sra** [comment]
Update: We tested with the fix the FLASH messages and are still seeing the issue.

**Comment by edenhill** [comment]
I belive this is fixed now by 4a9891b

**Comment by edenhill** [comment]
This issue is now fixed on master, please try to verify the fix in your environment.
Thank you.
